README.md

```diff
@@ -133,10 +133,222 @@ Tests may be run individually (use `--help` for usage), or collectively via the
 ## Tests
 
 ### heap
+
+In x86 architecture, the stack grows up and towards the heap (which by convention grows in the opposite direction to the stack).
+![Heap and stack](heap-stack.png)
+
+By the sample /proc/<pid>/maps listing provided at the top, the stack is currently set at address range, 7fffd27c3000-7fffd27e400. Thus it only takes 47 bits (log 7fffd27e4000 / log 2) to address this space as the heap can't grow past the stack. Hence,currently, only 47 bits are needed to address the heap even while using the 5-level page tables.
+
+This program tests to check if heap allocations can surpass 4-level page table tests.
+
 #### malloc
+
+The malloc test allocates as much memory as it can, 1 GiB at a time, pushing the heap up to higher and higher address ranges. (This memory is uninitialized.)
+Once the `malloc()` call fails, the test then returns the number of gigabytes allocated.
+
+Expected return values:
+```
+4-level page tables: ~130000 GiB
+5-level page tables: ~130000 GiB
+```
+as it only takes 47bits to address ~130000 GiB
+
 #### sbrk
-### mmap+memset+fork.c
+
+The sbrk test grows the heap by changing the location of the program break, which defines the end of a process's data segment. Like the malloc test, this effectively allocates more virtual memory space to the calling process.
+
+Expected return values:
+```
+4-level page tables: ~130000 GiB
+5-level page tables: ~130000 GiB
+```
+
+### mmap+memset+fork
+
+On x86 architecture, 5-level paging enables 56-bit userspace virtual address
+space. However, not all user space is ready to handle such wide addresses. To
+mitigate this, the kernel will not allocate virtual space above 47-bit by
+default.But userspace can ask for allocation from full address space by
+specifying hint address (with or without `MAP_FIXED`) above 47-bits.
+
+This program verifies that `mmap()` is possible beyond the 4-level
+page table range when supported and specifically requested.  The test
+program also verifies COW (copy on write) behaviour functions correctly
+while forking and sharing mappings.
+
+#### MAP_FIXED
+
+In order to successfully allocate large virtual address space as described by
+the Linux documentation
+(https://github.com/torvalds/linux/blob/master/Documentation/x86/x86_64/5level-paging.txt),
+the mmap+memset+fork will always call `mmap()` with the `MAP_FIXED` flag set.
```

Comment:
markdown syntax is `[linked words](http://...)`  so the "documentation" needs to be in square brackets.